XD-3571 Add Cassandra sink module

trisberg opened the pull request:
- Ported existing Spring XD Cassandra sink module

trisberg committed:
XD-3571 Add Cassandra sink module

- Ported existing Spring XD Cassandra sink module

artembilan commented on cassandra-sink/pom.xml:
```diff
@@ -0,0 +1,58 @@
+		<dependency>
+			<groupId>io.projectreactor</groupId>
+			<artifactId>reactor-core</artifactId>
+			<scope>test</scope>
```
I wonder if we can get rid of this dependency for that tiny `Supplier` usage and just switch to Java 8 for tests.

ericbottard replied:
Could even port our EventuallyMatcher from XD, which is more versatile than just `assertEqualsEventually`

ericbottard replied:
Hmm, on second thoughts, not sure it works exactly the same way. You need repetitive evaluation of the template here.
In all cases, dragging reactor for just a single callback interface indeed feels overhead.

artembilan replied:
When I was introducing that `assert` for the XD Modules project the Reactor dependency was there automatically from some XD dependency.
Since we have to specify it here explicitly, I'm sure we can remove it in favor of some other solution. And even if we can't use Java 8 from here, we can just introduce that simple `Supplier` interface as an inner class definition exactly in that specific test class.

ericbottard replied:
Absolutely

trisberg replied:
Agree, I'm adding an inner class interface def and removing reactor dependency

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraConfiguration.java:
```diff
@@ -0,0 +1,139 @@
+public class CassandraConfiguration extends AbstractCassandraConfiguration {
+
+	@Autowired
+	CassandraProperties cassandraSinkProperties;
```
Bad name for the property.
My idea was to extract some common options when we will add the Cassandra Source module.
That's why we have that specific `CassandraSinkProperties`.

trisberg replied:
Yes, will rename it

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkConfiguration.java:
```diff
@@ -0,0 +1,205 @@
+ */
+@Configuration
+@Import(CassandraConfiguration.class)
+@EnableIntegration
```
Having `@SpringBootApplication` this is redundant.
Minor, of course...

trisberg replied:
Yep, less code is better

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkConfiguration.java:
```diff
@@ -0,0 +1,205 @@
+	private CassandraOperations template;
+
+	@Bean
+	public MessageChannel input() {
```
This isn't in use any more.

trisberg replied:
Right, forgot to remove it

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSink.java:
```diff
@@ -0,0 +1,42 @@
+
+	@ServiceActivator(inputChannel=Sink.INPUT)
+	public void cassandraSink(Message<?> message) {
+		cassandraSinkMessageHandler.handleMessage(message);
```
Looking to this class and how it is wired with the `CassandraSinkConfiguration`, I'd call it as an overhead.
How about rename the `CassandraSinkConfiguration` to this name, move `@EnableBinding(Sink.class)` there, as well as this `@ServiceActivator` annotation?
Otherwise we end up with extra `handleMessage()`...

markfisher replied:
+1 but we can even keep the name `CassandraSinkconfiguration` since there is no requirement to have a class that is _the_ Sink.

artembilan commented on cassandra-sink/src/test/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkIntegrationTests.java:
```diff
@@ -0,0 +1,200 @@
+    @Autowired
+    ConfigurableApplicationContext applicationContext;
+
+	private static Cluster cluster;
```
One of us has wrong indents confing...
Tabs instead of whitespaces. 
Or have I missed anything already?..

trisberg replied:
must have cut'n pasted this code, should be tabs, fixing

artembilan commented on cassandra-sink/src/test/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkIntegrationTests.java:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	@WebIntegrationTest({"spring.cassandra.port=9043", 
+		"spring.cassandra.keyspace=cassandraTest",
```
Can't we really use here constants defined above ?

ericbottard replied:
Can.

Can't use a dynamic port though :(

trisberg replied:
switching to use constants. For dynamic ports, not even sure that is possible with Cassandra, There are quite a few ports required in the yml file, not sure if we could set system properties at  runtime to override them. I did change a few unneeded ones to use '0'.

artembilan commented on cassandra-sink/src/test/resources/cassandra.properties:
```diff
@@ -0,0 +1,4 @@
+info.shortDescription = Ingests data into a Cassandra table.
+options_class = org.springframework.xd.cassandra.CassandraSinkOptionsMetadata
+base_packages = org.springframework.xd.cassandra
```
What is the purpose of this `cassandra.properties` file, especially in the `test` scope ?

ericbottard replied:
This is XD stuff, needs to go.

trisberg replied:
Oops, got copied over, removing.

ericbottard commented on cassandra-sink/pom.xml:
```diff
@@ -0,0 +1,58 @@
+					<artifactId>slf4j-log4j12</artifactId>
+				</exclusion>
+			</exclusions>
+			<scope>test</scope>
```
Maybe move this scope element before the exclusions.
Also, a comment explaining why there are exclusions would be welcome

trisberg replied:
Sure, @artembilan any notes on why these exclusions are necessary, I'm assuming it is because the `cassandra-unit-spring` dependency is older and bring sin outdated artifacts plus the usual logging problems.

artembilan replied:
Yes, that's the best reason.
And maybe that's why guys don't comment it even on the core project: https://github.com/spring-projects/spring-data-cassandra/blob/master/pom.xml#L149

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraConfiguration.java:
```diff
@@ -0,0 +1,139 @@
+	@PostConstruct
+	public void init() throws IOException {
+		if (cassandraSinkProperties.getInitScript() != null) {
+			Resource initScriptResource = new DefaultResourceLoader().getResource(cassandraSinkProperties.getInitScript());
```
Can make the property on the `cassandraSinkProperties` object a `Resource` directly

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraConfiguration.java:
```diff
@@ -0,0 +1,139 @@
+			}
+			finally {
+				if (scanner != null) {
+					scanner.close();
```
try with resource is your friend here :)

artembilan replied:
???
Not sure what you mean...
I even don't understand right now why I use here `scanner`...
But anyway I don't see difference.
We read the whole file to string and then delimit it to several scripts.

ericbottard replied:
I'm saying you can do 

```
try (Scanner scanner = new Scanner(...)) {
   scripts = scanner.useDelimiter("\\A").next();
}
```

artembilan replied:
Doh... Haven't paid attention to the Java 7 features yet...
Thank you ! :smile:

trisberg replied:
making this change

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraProperties.java:
```diff
@@ -0,0 +1,162 @@
+public class CassandraProperties {
+
+	/*
+	 * The comma-delimited string of the hosts to connect to Cassandra
```
All javadoc comments need dots at the end.

ericbottard replied:
And need to be actual javadoc, not regular comments

trisberg replied:
fixing

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraProperties.java:
```diff
@@ -0,0 +1,162 @@
+		return this.contactPoints;
+	}
+
+	public int getPort() {
```
can add range 0-65535 validation

trisberg replied:
yep, adding

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkConfiguration.java:
```diff
@@ -0,0 +1,205 @@
+			cassandraMessageHandler.setIngestQuery(cassandraSinkProperties.getIngestQuery());
+		}
+		else if (StringUtils.hasText(cassandraSinkProperties.getStatementExpression())) {
+			Expression expression = PARSER.parseExpression(cassandraSinkProperties.getStatementExpression());
```
Can make the field on the properties object an `Expression` directly, provided you add `SpelExpressionConverterConfiguration` see https://github.com/spring-cloud/spring-cloud-stream-modules/blob/b12eaba454a202a91668ee4dd4f7518bebed3616/transform-processor/src/main/java/org/springframework/cloud/stream/module/transform/TransformProcessor.java#L35

artembilan replied:
Didn't we find the compromise with this `Expression` stuff ?
At least, shouldn't we follow here with the `FileSinkConfiguration` experience ?

artembilan replied:
This isn't addressed yet and even without comments...

trisberg replied:
pushed new fix to address this

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkConfiguration.java:
```diff
@@ -0,0 +1,205 @@
+	private static class ISO8601StdDateFormat extends StdDateFormat {
+
+		@Override
+		protected boolean looksLikeISO8601(String dateStr) {
```
What's the point of this?

artembilan replied:
???
Please, be more specific.
1. 

```
/**
  * Defines a commonly used date format that conforms
  * to ISO-8601 date formatting standard, when it includes basic undecorated
  * timezone definition
*/
protected final static String DATE_FORMAT_STR_ISO8601 = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
```

You can kill me, but I don't remember already why I have chosen this format...
Maybe it is some String format which can be inserted as is to the Cassandra Date column?..
http://docs.datastax.com/en/cql/3.0/cql/cql_reference/timestamp_type_r.html 
1. We need an extension for this class to have access to this `boolean` method. See the logic in the `transformPayload()` above.

ericbottard replied:
My point was: why override a method just to call `super` ?

ericbottard replied:
> We need an extension for this class to have access to this boolean method. See the logic in the transformPayload() above.

Ah, I think I see. This makes the method visible. Maybe we should add a comment to explain this then

trisberg replied:
adding a comment

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkConfiguration.java:
```diff
@@ -0,0 +1,205 @@
+						if (value instanceof String) {
+							String string = (String) value;
+							if (this.dateFormat.looksLikeISO8601(string)) {
+								synchronized (this.dateFormat) {
```
why not instantiate one per method invocation really?

artembilan replied:
Let me guess: because GC is more expensive than short `synchronized`.
Pay attention that its field is `final`, so it isn't counted by GC at all.

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkProperties.java:
```diff
@@ -0,0 +1,109 @@
+public class CassandraSinkProperties {
+
+	/*
+	 * The consistencyLevel option of WriteOptions
```
Dots at end, need real javadoc style comments (not regular comments)

trisberg replied:
fixing

ericbottard commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkProperties.java:
```diff
@@ -0,0 +1,109 @@
+	private String ingestQuery;
+
+	/*
+	 * The expression in Cassandra query DSL style
```
Says "in Cassandra query DSL style", yet this is parsed as SpEL o.O

See my earlier comment about making this a SpEL `Expression` object directly then, if applicable

trisberg replied:
Are they mutually exclusive?

ericbottard commented on cassandra-sink/src/test/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkIntegrationTests.java:
```diff
@@ -0,0 +1,200 @@
+
+	private static final String CASSANDRA_STORAGE_CONFIG = "spring-cassandra.yaml";
+
+	private static final int PORT = 9043; // See spring-cassandra.yaml - native_transport_port
```
This should be made dynamic and find an available port

ericbottard commented on cassandra-sink/src/test/resources/int-db.cql:
```diff
@@ -0,0 +1,11 @@
+DROP TABLE IF EXISTS book;
```
should this be named in_i_t-db.cql ?

artembilan replied:
+1.
Although it doesn't matter because it is the test scope resource.

ericbottard commented on cassandra-sink/src/test/resources/logback.xml:
```diff
@@ -0,0 +1,10 @@
+<configuration>
```
this should be removed

trisberg replied:
gone

artembilan replied:
??? hasn't gone yet...

trisberg replied:
it was gone for a while, added it back in to get rid of all the logging during @BeforeClass

artembilan replied:
Yes, that's why I ask in the `s-c-s` Slack channel:

> @here, can you tell me where is the `logging` config for the `s-c-s-modules` and why so much `DEBUG` noise do you allow? Thanks

trisberg committed:
XD-3571 Add Cassandra sink module

- Addressing review comments

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkProperties.java:
```diff
@@ -0,0 +1,109 @@
+	/**
+	 * The expression in Cassandra query DSL style.
+	 */
+	private String statementExpression;
```
> Says "in Cassandra query DSL style", yet this is parsed as SpEL o.O

My bad. Not sure how much letters you allow for such a comment on properties, but my intention was to say here that is:
A SpEL to be evaluated to the `com.datastax.driver.core.Statement` where Cassandra query DSL can be used without FQCN. E.g. 

```
--statementExpression=QueryBuilder.select('c1','c2').from('t').where(QueryBuilder.eq('c3', payload))
```

> Are they mutually exclusive?

If you mean together with `ingestQuery`, than yes. they are. But the `ingestQuery` has precedence in the `CassandraMessageHandler`. All other options are just ignored.

artembilan replied:
If you'd like to see some other comment on this property, please, count my comment here.

trisberg commented:
@artembilan I was going to update this PR but the integration tests no longer pass. Get this error - com.datastax.driver.core.exceptions.InvalidQueryException: Keyspace 'cassandraTest' does not exist. To me it looks like we create that when the test class is loaded. Any ideas?

artembilan commented:
@trisberg , would you mind sharing your code to play from my side locally?

Thanks

artembilan commented:
Well, it doesn't pass for me even without your polishing.
Investigating....

artembilan commented:
Not sure what have been changed, but it can't connect because the `keyspace` must be quoted:
http://log.damnpeople.fr/2014/10/cassandra-keyspace-case-sensitivity-wtf/
http://stackoverflow.com/questions/29569443/cassandra-keyspace-name-with-hypen

So, we should just fix our `cassandraTest` name to something like `cassandratest` or just `sink`.
But... That doesn't help me to pass the test. I'm finishing with something like JVM crash. 
Looks like anything has been changed recently in the SF which tells us that we aren't correct.

artembilan commented on cassandra-sink/src/test/resources/spring-cassandra.yaml:
```diff
@@ -0,0 +1,630 @@
+# will spread data evenly across them, subject to the granularity of
+# the configured compaction strategy.
+data_file_directories:
+    - build/embeddedCassandra/data
```
s/build/target
just because we are with Maven now, not Gradle...

artembilan commented on cassandra-sink/src/test/resources/spring-cassandra.yaml:
```diff
@@ -0,0 +1,630 @@
+    - build/embeddedCassandra/data
+
+# commit log
+commitlog_directory: build/embeddedCassandra/commitlog
```
DITO

artembilan commented on cassandra-sink/src/test/resources/spring-cassandra.yaml:
```diff
@@ -0,0 +1,630 @@
+# row_cache_keys_to_save: 100
+
+# saved caches
+saved_caches_directory: build/embeddedCassandra/saved_caches
```
DITO

artembilan commented on cassandra-sink/src/test/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkIntegrationTests.java:
```diff
@@ -0,0 +1,202 @@
+@DirtiesContext
+public abstract class CassandraSinkIntegrationTests {
+
+	public static final String CASSANDRA_KEYSPACE = "cassandraTest";
```
So, as we talked before this should be changed to something only in lower case.
Just because CQL doesn't care about quotes.

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraConfiguration.java:
```diff
@@ -0,0 +1,133 @@
+	private Session session;
+
+	@PostConstruct
+	public void init() throws IOException {
```
Looks like we aren't correct here and introduce the circular dependency:
`session` is created by the `@Bean CassandraSessionFactoryBean` in the superclass.

Right, when I comment `@Autowired` on the `session` and `@PostConstruct` the test goes a bit further.

Investigating what's going on here with `Invalid version for TimeUUID type`.

Meanwhile I'd ask you for the best place to call init scripts for the DB before the start of our adapters, of course...

artembilan replied:
I hacked it with moving all this initialization to the inner class:

```
@Configuration
protected static class CassandraInitializerConfiguration {

 }
```

After that it started to pass for me - the whole `cassandra-sink` module.
It is very long testing process though with a lot of `DEBUG` in logs, but it passes.

So, @trisberg , feel free to push your changes and than I'll share with you my polishing. 
Or you can address it yourself already, of course.

trisberg committed:
XD-3571 Add Cassandra sink module

- Addressing additional review comments

- Removing CassandraSink and moving @ServiceActivator to CassandraSinkConfiguration class

- Removing CassandraConfiguration init logic, passing init scripts via `getStartupScripts()` method

- Changing CassandraEntityInsertTests to use UUIDs.timeBased() for the UUID value

trisberg commented:
Think I solved the test issues. Rebased the code.

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraProperties.java:
```diff
@@ -0,0 +1,164 @@
+	/**
+	 * The keyspace name to connect to.
+	 */
+	@Value("${spring.application.name:tmp}")
```
Is the `tmp` intentional?

trisberg replied:
yes, other sinks tend yo use tmp or /tmp as a default

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkApplication.java:
```diff
@@ -0,0 +1,31 @@
+ * @author Thomas Risberg
+ */
+@SpringBootApplication
+public class CassandraSinkApplication {
```
It's interesting if we need this separate class at all.
Looks like it is the same in each module, which are the independent applications at all. 
So, no need to worry about the name of class.

We even could generate it... Or try to figure out some other way to bypass this boring copy/paste work...

artembilan commented on cassandra-sink/src/test/java/org/springframework/cloud/stream/module/cassandra/sink/CassandraSinkIntegrationTests.java:
```diff
@@ -0,0 +1,204 @@
+		Book b;
+		for (int i = 0; i < numBooks; i++) {
+			b = new Book();
+			b.setIsbn(UUID.randomUUID());
```
What is interesting that this test data is the same like for the `insert` test.
From other side we have the table script:

```
isbn        uuid PRIMARY KEY,
```

not `timeuuid`, like it barks to us for the `insert` test...

artembilan replied:
I know "why?": you had `"spring.cassandra.init-script=init-db.cql` before for this test, too.
But right now it is based on the Spring Data Entity Mapping mechanism.
And we end up with the

```
TIMEUUID  (15, UUID.class),
```

for `Book.isbn`.

So, that your `book.setIsbn(UUIDs.timeBased());` is a good fix.

artembilan commented:
The top comment.
You should configure your IDE to remove trailing whitespaces. Otherwise a lot of auto-polishing from my side :smile:

trisberg commented:
I do have that configured, but only for "Modified Lines", not sure where they came from, maybe the original files I copied.

artembilan commented:
Well, I use "All" mode to fix an existing code during PRs.

trisberg committed:
XD-3571 Add Cassandra sink module

- Addressing white space issue

- Removing extra @DirtiesContext

trisberg committed:
XD-3571 Add Cassandra sink module

- Adding log configuration for tests

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraConfiguration.java:
```diff
@@ -0,0 +1,134 @@
+	CassandraProperties cassandraProperties;
+
+	@Override
+	protected List<String> getStartupScripts() {
```
Let me replicate my comment from commit here one more time.

I didn't use this approach from the beginning because this getStartupScripts() is used for Cluster configuration as well as for Session.
See the usage of this method.

My intention of these init scripts was only to initialize the specific keyspace. And only session is for that.
I admit that it looks like a bug in the Spring Data Cassandra, because we are sharing the same scripts for the whole Cluster and the specific session for particular keyspace.
Maybe I don't know Cassandra so well, but I'm afraid that we may have some side effect to go this strange way...

artembilan replied:
As I said it is dangerous. Scripts are executed on the `system` keyspace - `CassandraCqlClusterFactoryBean`:

```
if (system == null) {
                    system = scripts.size() == 0 ? null : cluster.connect();
                }

                if (template == null) {
                    template = system == null ? null : new CqlTemplate(system);
                }

                for (String script : scripts) {

                    if (log.isDebugEnabled()) {
                        log.debug("executing raw CQL [{}]", script);
                    }

                    template.execute(script);
}
```

So, in our case we create the same table in two keyspaces.
Fortunately with Embedded Cassandra we have all those rights.
With the real Cassandra need to test.
But, at least, I sent you a bell here...

artembilan replied:
I have it locally with inner class like this:

``` java
public class CassandraConfiguration extends AbstractCassandraConfiguration {

    @Autowired
    CassandraProperties cassandraProperties;

    /**
     * The inner class to execute init scripts on the provided {@code keyspace}.
     * It is here to bypass circular dependency with {@link Session} injection and
     * its {@code @amp;Bean} in the {@link AbstractCassandraConfiguration}.
     * From other side we can't use {@link #getStartupScripts()} because they are
     * shared between Cluster and Session initializations.
     */
    @Configuration
    protected static class CassandraInitializerConfiguration {

        @Autowired
        CassandraProperties cassandraProperties;

        @Autowired
        private Session session;

        @PostConstruct
        public void init() throws IOException {
            if (cassandraProperties.getInitScript() != null) {
                Resource initScriptResource = new DefaultResourceLoader().getResource(cassandraProperties.getInitScript());
.....
```

trisberg committed:
XD-3571 Add Cassandra sink module

- Moving init logic back to CassandraConfiguration `@PostConstruct` using inner class to avoid circular reference of `session` and not using `getStartupScripts()` method since that is cluster wide and not keyspace specific

artembilan commented:
Everything else for me looks good here.

trisberg committed:
XD-3571 Add Cassandra sink module

- Changing statementExpression property to be type `Expression`

trisberg committed:
XD-3571 Add Cassandra sink module

- Addressing trailing white space issues

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraProperties.java:
```diff
@@ -0,0 +1,164 @@
+import org.springframework.util.StringUtils;
+
+/**
+ * Properties for the cassandra sink.
```
This is common properties. Not only for the `sink`.
Minor: will be fixed on merge.

artembilan commented on cassandra-sink/src/main/java/org/springframework/cloud/stream/module/cassandra/CassandraProperties.java:
```diff
@@ -0,0 +1,164 @@
+	/**
+	 * The path to file with CQL scripts (delimited by ';') to initialize keyspace schema.
+	 */
+	private String initScript;
```
I found that this one can be specified as a `Resource`.
Then in the debug I see this result:

```
result = {DefaultResourceLoader$ClassPathContextResource@7696} "class path resource [init-db.cql]"
 path = "init-db.cql"
 classLoader = {Launcher$AppClassLoader@7699} 
 clazz = null
```

Let me know if we can go ahead with the `Resource` approach and I'll fix it on merge.
Otherwise explain, please, what is the concern exists do not do that.
Thanks

ericbottard replied:
I'd vote for `Resource` type as well, and let the converter do the resolve

trisberg replied:
Resource sounds good

artembilan commented:
Here is a polishing for your confirmation: https://github.com/artembilan/spring-cloud-stream-modules/commit/b1404ac688ba26d225d55e9c19f9418ebb2ae782.
Tested against "real" Cassandra, although it doesn't matter: the Cassandra is just a Java application :smile:

trisberg commented:
Looks OK. One comment, it seems you use the `this.` prefix everywhere while most other modules do not. Seems to be an SI thing, don't think most other Spring projects use that style.

artembilan commented:
No, that is really SF code style on which we rely in SI: https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style:

> A field of a class should always be referenced using this. A method of class, however, should never be referenced using this.

artembilan commented:
Merged as https://github.com/spring-cloud/spring-cloud-stream-modules/commit/29779f033bd38d62cbe5e5161621cff7af8170cf
